packages/unified-selection/src/unified-selection/HiliteSetProvider.ts

```diff
@@ -0,0 +1,338 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+
+/** @packageDocumentation
+ * @module UnifiedSelection
+ */
+
+import { ECClass, IMetadataProvider, parseFullClassName } from "./queries/ECMetadata";
+import { ECSqlBinding, IECSqlQueryExecutor } from "./queries/ECSqlCore";
+import { SelectableInstanceKey, Selectables } from "./Selectable";
+
+/**
+ * A set of model, subcategory and element ids that can be used for specifying hilite.
+ * @public
+ */
+export interface HiliteSet {
+  models: string[];
+  subCategories: string[];
+  elements: string[];
+}
+
+/**
+ * Props for creating a `HiliteSetProvider` instance.
+ * @public
+ */
+export interface HiliteSetProviderProps {
+  queryExecutor: IECSqlQueryExecutor;
+  metadataProvider: IMetadataProvider;
+}
+
+/**
+ * ECSQL based provider which determines what `HiliteSet`
+ * should be hilited based on the supplied `Selectables`.
+ * @public
+ */
+export class HiliteSetProvider {
+  private _queryExecutor: IECSqlQueryExecutor;
+  private _metadataProvider: IMetadataProvider;
+  // Map between a class name and its type
+  private _classRelationCache: Map<string, InstanceIdType>;
+  private _hasFunctionalSchema?: boolean;
+
+  private constructor(props: HiliteSetProviderProps) {
+    this._queryExecutor = props.queryExecutor;
+    this._metadataProvider = props.metadataProvider;
+    this._classRelationCache = new Map<string, InstanceIdType>();
+  }
+
+  /**
+   * Create a hilite set provider.
+   */
+  public static create(props: HiliteSetProviderProps) {
+    return new HiliteSetProvider(props);
+  }
+
+  /**
+   * Get hilite set for supplied `Selectables`.
+   */
+  public async getHiliteSet(selection: Selectables): Promise<HiliteSet> {
+    // Map between element ID types and EC instance IDs
+    const keysByType = new Map<InstanceIdType, string[]>();
+    if (!this._hasFunctionalSchema) {
```

Comment:
shouldn't this check for `undefined` instead?

Reply:
You are right, fixed.